### ISSUE TITLE
hashsum_files could include the filename

### ISSUE THREAD

**Issue by manthey** [girder/girder#2665]
The sha512sum CLI will locate a file and validate it if the hashsum file is of the form `<hash>  <file name>\n`.  We emit hashsum files with just `<hash>\n`, which is what CMake's external data consumes.  We should consider adding an option to generate the hashsum file which includes the filename as an option.

This could be exposed in the UI with a second key icon with a tooltip to explain the difference, but there might be a better way.

**Comment by brianhelba** [comment]
+1 to this feature, as the integration with standard tools like `sha512sum` would be great.

I'd like Girder to be parsimonious, and avoid options for returning the same data in slightly-different formats, particularly in light of the [Robustness principle](https://en.wikipedia.org/wiki/Robustness_principle). Ultimately, I think it might be cleaner to update Girder's returned hashsum file format as part of a major release.

Of course, we don't want to break CMake ExternalData. So, I think a tractable, positive step for right now would be to update CMake ExternalData to be more permissive in the keyfile format that it accepts. This would benefit both CMake and Girder regardless of the path forward.

**Comment by johnkit** [comment]
+1. The standard/expected file format is `<hash>  <file>`, so people who want to use the hash to check their download have to manually construct the a file. 

Regarding the UI, are there other end-user reasons for downloading the sha file besides checking a download? If not, I would vote to replace the current icon with a (new) function to download the sha file in standard format (i.e., sha plus filename).

**Comment by zachmullen** [comment]
> Regarding the UI, are there other end-user reasons for downloading the sha file besides checking a download?

The primary use case for this endpoint isn't to check a file hash, it was developed for use within ExternalData, so that small plaintext references to large blob data could be stored inside version control systems. Using these files to check hashsums wasn't really part of the design, though it would be nice for that to "just work" with common hash check utilities.

We also have clients of this plugin besides just CMake ExternalData, for instance there is a downstream that uses these key files within a bazel build system. I would object to breaking the current default behavior *even as part of a major release*, as URL stability is so fundamentally important to the concept of ExternalData, regardless of what client is used -- the point is that someone can go back to any commit in their history and run a script, and the correct test data from that point in time will be used. So I would only be OK with a new format if we preserve the existing endpoint behavior & format.

(As an aside, I don't understand why we are opposed to serving data in different formats based on request options, it seems like a fairly standard practice, see e.g. the HTTP `Accept` header behavior. I get that we should avoid it generally, but I don't think we should be afraid to support such functionality where it's pragmatic.)

**Comment by zachmullen** [comment]
Adding some of the power users of this plugin + ExternalData so they have a chance to weigh in:

@thewtex @jcfr

**Comment by johnkit** [comment]
1. We certainly should not break the current endpoint.

2. In terms of UI: As an _end-user_ of ExternalData, I have never had to go into Girder and click that icon. But if the current icon is needed to facilitate developers in adding or modifying ExternalData, then sure, I would change my vote to have 2 icons.

**Comment by jcfr** [comment]
> I would only be OK with a new format if we preserve the existing endpoint behavior & format.

Definitively

> idea 1: with a second key icon with a tooltip to explain the difference,
> idea 2: would vote to replace the current icon with a (new) function to download the sha file in standard format

Generally since we can agree the idea to minimize surprise effect and avoid confusion while helping users. 

Here is a possible plan:

After the improved the plugin is deployed and the first time the user click on the key icon, we could pop-up an explanation where we could ask him to select his/her preferred format for key file.

Preferring the new format would simply mean the existing endpoint would be given a new parameter.

>  CMake ExternalData to be more permissive in the keyfile format that it accepts

:+1: 

> hashsum_files could include the filename

Now, considering that any given hash can be associated with one or more files. The file will have to be generated within a "context".

Context could be:

* user click on the UI to get info for as specific file. Format is as described above `<hash> <file>`

* end-point is called without a specific context or by specifying a collection, folder or item. In that case, the file would contain the list like this one:

```
<hash> <file1>
<hash> <file2>
...
```

To summarize, for all reason listed above:
* we need to keep the existing end point work the way it is.
* we could add a new parameter (e.g `context`) to the endpoint. Default value would be `None` with possible values being [None, `True`, `<id_of_resource>`]
* we could improve the UI by asking user his preferred format the first time they download the key file

**Comment by thewtex** [comment]
+1 to @zachmullen 's comments. The existing behavior (the content of the file, and the link to the keyfile) need to keep working the same way to avoid breaking ExternalData in ITK. We use the current UI:

  https://itk.org/ITKExamples/Documentation/Contribute/UploadBinaryData.html
